File: 프로그래머스/lv3/43162. 네트워크/네트워크.py
from collections import deque

def solution(n, computers):
    answer  = 1
    queue = deque([0])
    visited = [0] * n
    visited[0] = 1
    
    while 0 in visited:
        node = queue.popleft()
        for idx in range(len(computers[node])):
            if computers[node][idx] == 1 and visited[idx] == 0: 
                queue.append(idx)  
                visited[idx] = 1
        # 네트워크 종료 check
        if len(queue) == 0:
            answer += 1
            for i in range(len(visited)):
                if visited[i] == 0:
                    queue.append(i)
                    break
    return answer
            
        
            
            
            
    
    

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;
class Solution {
    public int solution(int n, int[][] computers) {
        int answer = 0;
        boolean[] visited = new boolean[n];
        
        for(int i = 0; i < n; i++){
            //방문하지 않았다면 방문해서 dfs 돌리고 판단
            if(!visited[i]){
                dfs(i, computers, visited, n);
                answer++;
            }
        }
        return answer;
    }
    
    private void dfs(int visitedIndex, int[][] computers, boolean[] visited, int n){
        visited[visitedIndex] = true;
        for(int i = 0; i < n; i++){
           if(computers[visitedIndex][i] == 1 && visited[i] == false){
               dfs(i, computers, visited, n);
           }
        }
    }
}